### PR TITLE
Jon/aave/loan-status-redesign

### PR DESCRIPTION
**Pull request by Jon-edge**
This PR updates the Action display implementation for Loan Create workflows to support the new spec requirement of supporting multiple display string options for the same ActionOp type

Simulator Screen Shot - iPhone 8 ios 16 - 2022-10-24 at 19 49 00

#### PR Dependencies

 none

#### PR Requirements

If you have made **any** visual changes to the GUI. Make sure you have:
- [x] Tested on iOS device
- [ ] Tested on Android device
- [ ] Tested on small-screen device (iPod Touch)
- [ ] Tested on large-screen device (tablet)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203187813657970

### CODE REVIEW

File: src/components/scenes/Loans/LoanCreateConfirmationScene.tsx
Comment (samholmes) [inline src/components/scenes/Loans/LoanCreateConfirmationScene.tsx:110]: Swaps can take a very long time to complete. If we cover 2x we're ensuring the product works even if the fees double during the time we're waiting on swaps. If they don't double, then the user has a bit more ETH left over; so what, that can be used for future contract calls to manage their loan. Much better to overshoot and have extra ETH then to undershoot and fail the program asking the user to recover the broken loan state, don't you think?

**Review by samholmes — CHANGES_REQUESTED** [review]
Looks like what's missing in this change is persisting the new display data to disk. The cleaners haven't been touched.

The change does mix concerns (display data with program data), which still irks me. I'd be much happier with the display data being separate from the program or ActionOp object.

One idea I had (but had not really mentioned because I didn't want to bike-shed), was to simply add a `programIntent` field to the `ActionProgram`. The display implementation can use the `programIntent` to determine how to translate the `(ActionProgram, ActionProgramState)` data into `ActionProgramDsiplayInfo` or any other dynamic type; we can completely re-write the display API at any time. The only think we're committing to as frozen API is a single intent field and even if we deprecate a value for this (or mess up usage), we can always fallback to some generic display (keeping our guarantee that the set of all action programs is included in the domain for the `getActionProgramDisplayInfo` function).

File: src/components/scenes/Loans/LoanCreateConfirmationScene.tsx
Comment (Jon-edge) [reply src/components/scenes/Loans/LoanCreateConfirmationScene.tsx:110]: That's fair, I did not consider the possible length of time. However, if I didn't consider it, another user wouldn't either, and would be understandably confused about this.

File: src/controllers/action-queue/types.ts
Comment (samholmes) [inline src/controllers/action-queue/types.ts:116]: This can now be removed.

File: src/controllers/action-queue/types.ts
Comment (samholmes) [inline src/controllers/action-queue/types.ts:112]: I'd much rather put this action item in a task rather than a TODO comment.

File: src/controllers/action-queue/types.ts
Comment (samholmes) [inline src/controllers/action-queue/types.ts:85]: What if the type for `SwapActionOp['displayKey']` was different from the type for `SeqActionOp['displayKey']` and `ParActionOp['displayKey']`? The reason for this is to treat the keys for a particular action op distinctly from the other action ops. This way a display key for the SwapActionOp is specific to the `SwapActionOp`. This can avoid _some of_ the runtime type error problems we could run into if a key is used on an ActionOp that isn't not suppose to be used on.

Here's an example of what I mean:

```ts
export type SwapActionOp = {
  type: 'swap'
  amountFor: 'from' | 'to'
  fromTokenId?: string
  fromWalletId: string
  nativeAmount: string
  toTokenId?: string
  toWalletId: string
  displayKey?: SwapActionOpDisplayKey
}
export type SwapActionOpDisplayKey = 'swap-deposit' | 'swap-deposit-fees'
```

And here's an example for what we can do with `ParActionOp`:

```ts
export type ParActionOp = {
  type: 'par'
  actions: ActionOp[]
  displayKey?: ParActionOpDisplayKey
}
export type ParActionOpDisplayKey = 'create'
```

Making the set of valid keys specific to the ActionOp avoid the category of error of putting say `swap-deposit` key on ` ParActionOp`. Also it make it very clear to the developer about which keys apply to which `ActionOp`; self-documenting code.

File: src/components/scenes/Loans/LoanCreateConfirmationScene.tsx
Comment (samholmes) [inline src/components/scenes/Loans/LoanCreateConfirmationScene.tsx:108]: Can we hold off on this change just because we don't want it impacting our launch. We can create a task to reconsider this post-beta?

File: src/controllers/action-queue/cleaners.ts
Comment (samholmes) [inline src/controllers/action-queue/cleaners.ts:76]: Just a trick to use for if we were to have distinct `*ActionOpDisplayKey` types (my final ask in this review):

```ts
const asParActionOp = asObject<ParActionOp>({
  type: asValue('par'),
  actions: asArray((raw: any) => asActionOp(raw)),
  displayKey: (raw: unknown) => asOptional(asParActionOpDisplayKey)(raw)
})
const asParActionOpDisplayKey = asValue('create')
```

This extra anonymous arrow-function is how we can reference a cleaner before it's defined.

File: src/controllers/action-queue/ActionProgram.ts
Comment (samholmes) [inline src/controllers/action-queue/ActionProgram.ts:17]: This field isn't necessary here because it's injected by the ActionQueueStore

File: src/controllers/action-queue/display.ts
Comment (samholmes) [inline src/controllers/action-queue/display.ts:27]: Can't we support `LeafActionOp` types with the pure inference method?

**Review by samholmes — CHANGES_REQUESTED** [review]
Some minor changes, and one main ask. Feel free to rebase-autosquash the branch so far.

File: src/controllers/action-queue/display.ts
Comment (Jon-edge) [reply src/controllers/action-queue/display.ts:27]: Ah yeah that was the point in the first place, good catch

File: src/controllers/action-queue/ActionProgram.ts
Comment (Jon-edge) [reply src/controllers/action-queue/ActionProgram.ts:17]: I like the redundancy regardless, for debugging prior to the ActionQueueStore or if implementation changes

File: src/controllers/action-queue/display.ts
Comment (Jon-edge) [reply src/controllers/action-queue/display.ts:27]: Actually I did try that. Right now, we can't go by inference because getActionOpDisplayInfo drills down into the leaves. When we fully isolate the display implementation we can rely more on static typing

Comment (Jon-edge) [reply src/controllers/action-queue/display.ts:27]: Actually this check isn't needed if I instead restrict the ActionProgram's root actionOp type as a NodeActionOp

Comment (samholmes) [reply src/controllers/action-queue/display.ts]: I don't see why we would ever have a `LeafActionOp` in an action program... This is just one action which doesn't require the action queue, and even if we did want a program with one action for whatever reason, adding the `NodeActionOp` could just be required overhead. I'm okay with making this change as a breakaway task.

File: src/controllers/action-queue/types.ts
Comment (samholmes) [inline src/controllers/action-queue/types.ts:164]: Oh you did it. Awesome; cool cool

Comment (samholmes) [reply src/controllers/action-queue/types.ts]: Erm, needs revert cause it cascades type changes for the rest of the runtime implementation. Let's postpone this work item.

**Review by samholmes — APPROVED** [review]
Great job here! Approved.

Note, autosquash needed and perhaps double check this all works before merging.

**Pull request merged**
